### PR TITLE
Update README.md

### PR DESCRIPTION
**Pull request by NaderMalki**
import argparse
Bihemispheric neural network with 3D and bihemispheric processing Patent number: 140450140003002031
International:Inventor:Non-Property
Any use without permission from the owner of the invention will be subject to the legal license of the inventors' laws,
import random
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader, Subset
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix
import pandas as pd
import json
from pathlib import Path

# -------------------
# Synthetic Dataset
# -------------------
class SyntheticDualHemisphereDataset(Dataset):
    def __init__(self, n_samples=1000, seq_len=50, input_size=128, n_classes=2, seed=0):
        rng = np.random.RandomState(seed)
        self.X_left = rng.normal(size=(n_samples, seq_len, input_size)).astype(np.float32)
        self.X_right = rng.normal(size=(n_samples, seq_len, input_size)).astype(np.float32)
        sums = self.X_left.sum(axis=(1,2))
        thresh = np.median(sums)
        self.y = (sums > thresh).astype(np.int64)
        self.n_samples = n_samples

    def __len__(self):
        return self.n_samples

    def __getitem__(self, idx):
        return self.X_left[idx], self.X_right[idx], self.y[idx]

# -------------------
# Utilities
# -------------------
def set_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.use_deterministic_algorithms(False)

def stratified_split(dataset, val_frac=0.1, test_frac=0.1):
    ys = np.array([dataset[i][2] for i in range(len(dataset))])
    idxs_class0 = np.where(ys == 0)[0]
    idxs_class1 = np.where(ys == 1)[0]
    def split_indexes(idxs, val_frac, test_frac, rng):
        rng.shuffle(idxs)
        n = len(idxs)
        n_test = int(n * test_frac)
        n_val = int(n * val_frac)
        test_idx = idxs[:n_test]
        val_idx = idxs[n_test:n_test+n_val]
        train_idx = idxs[n_test+n_val:]
        return train_idx.tolist(), val_idx.tolist(), test_idx.tolist()
    rng = np.random.RandomState(0)
    t0, v0, s0 = split_indexes(idxs_class0.copy(), val_frac, test_frac, rng)
    t1, v1, s1 = split_indexes(idxs_class1.copy(), val_frac, test_frac, rng)
    train_idx = t0 + t1
    val_idx = v0 + v1
    test_idx = s0 + s1
    rng.shuffle(train_idx); rng.shuffle(val_idx); rng.shuffle(test_idx)
    return train_idx, val_idx, test_idx

def collate_batch(batch):
    lefts = torch.from_numpy(np.stack([b[0] for b in batch]))
    rights = torch.from_numpy(np.stack([b[1] for b in batch]))
    ys = torch.tensor([b[2] for b in batch], dtype=torch.long)
    return lefts, rights, ys

def evaluate(model, loader, device):
    model.eval()
    preds, trues = [], []
    with torch.no_grad():
        for xl, xr, y in loader:
            xl, xr = xl.to(device), xr.to(device)
            logits = model(xl, xr)
            pred = logits.argmax(dim=1).cpu().numpy()
            preds.extend(pred.tolist())
            trues.extend(y.numpy().tolist())
    acc = accuracy_score(trues, preds)
    f1 = f1_score(trues, preds, average='macro')
    cm = confusion_matrix(trues, preds)
    return acc, f1, cm

# -------------------
# Dummy Model (replace with your actual model)
# -------------------
class CoordinatedDualHemisphere(nn.Module):
    def __init__(self, input_size, hidden_size, num_layers, num_classes, bidirectional, dropout, fc_dropout):
        super().__init__()
        self.lstm_left = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True,
                                 bidirectional=bidirectional, dropout=dropout)
        self.lstm_right = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True,
                                  bidirectional=bidirectional, dropout=dropout)
        lstm_out_size = hidden_size * (2 if bidirectional else 1)
        self.fc = nn.Sequential(
            nn.Linear(2 * lstm_out_size, 128),
            nn.ReLU(),
            nn.Dropout(fc_dropout),
            nn.Linear(128, num_classes)
        )

    def forward(self, x_left, x_right):
        _, (h_left, _) = self.lstm_left(x_left)
        _, (h_right, _) = self.lstm_right(x_right)
        h_left = h_left[-1]
        h_right = h_right[-1]
        h = torch.cat([h_left, h_right], dim=1)
        return self.fc(h)

# -------------------
# Training Loop
# -------------------
def train_one_run(args, run_seed, out_dir):
    out_dir = Path(out_dir)
    out_dir.mkdir(parents=True, exist_ok=True)
    device = torch.device('cuda' if torch.cuda.is_available() and args.use_cuda else 'cpu')
    set_seed(run_seed)

    ds = SyntheticDualHemisphereDataset(n_samples=args.n_samples, seq_len=args.seq_len,
                                        input_size=args.input_size, seed=42)
    train_idx, val_idx, test_idx = stratified_split(ds, val_frac=args.val_frac, test_frac=args.test_frac)
    train_loader = DataLoader(Subset(ds, train_idx), batch_size=args.batch_size, shuffle=True, collate_fn=collate_batch)
    val_loader = DataLoader(Subset(ds, val_idx), batch_size=args.batch_size, shuffle=False, collate_fn=collate_batch)
    test_loader = DataLoader(Subset(ds, test_idx), batch_size=args.batch_size, shuffle=False, collate_fn=collate_batch)

    model = CoordinatedDualHemisphere(input_size=args.input_size, hidden_size=args.hidden_size,
                                      num_layers=args.num_layers, num_classes=args.n_classes,
                                      bidirectional=args.bidirectional, dropout=args.lstm_dropout,
                                      fc_dropout=args.fc_dropout).to(device)

    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=args.lr)
    scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, factor=0.5, patience=3)
    scaler = torch.cuda.amp.GradScaler() if args.use_amp and device.type == 'cuda' else None

    best_val_loss = float('inf')
    epochs_no_improve = 0
    best_checkpoint = None
    history = []

    for epoch in range(1, args.epochs + 1):
        model.train()
        epoch_losses = []
        for xl, xr, y in train_loader:
            xl, xr, y = xl.to(device), xr.to(device), y.to(device)
            optimizer.zero_grad()
            if scaler:
                with torch.cuda.amp.autocast():
                    logits = model(xl, xr)
                    loss = criterion(logits, y)
                scaler.scale(loss).backward()
                scaler.step(optimizer)
                scaler.update()
            else:
                logits = model(xl, xr)
                loss = criterion(logits, y)
                loss.backward()
                optimizer.step()
            epoch_losses.append(loss.item())

        val_loss = np.mean(epoch_losses)  # proxy
        val_acc, val_f1, val_cm = evaluate(model, val_loader, device)
        scheduler.step(val_loss)

        history.append({'epoch': epoch, 'train_loss': val_loss, 'val_acc': val_acc, 'val_f1': val_f1})
        if val_loss < best_val_loss - 1e-6:
            best_val_loss = val_loss
            epochs_no_improve = 0
            best_checkpoint = {'model_state': model.state_dict(), 'optimizer_state': optimizer.state_dict(), 'epoch': epoch}
            torch.save(best_checkpoint, out_dir / f'best_run_seed{run_seed}.pth')
        else:
            epochs_no_improve += 1
            if epochs_no_improve >= args.patience:
                break

    if best_checkpoint:
        model.load_state_dict(best_checkpoint['model_state'])
    test_acc, test_f1, test_cm = evaluate(model, test_loader, device)

    pd.DataFrame(history).to_csv(out_dir / f'history_seed{run_seed}.csv', index=False)
    with open(out_dir / f'metrics_seed{run_seed}.json', 'w') as f:
        json.dump({'test_acc': test_acc, 'test_f1': test_f1}, f)
    pd.DataFrame(test_cm).to_csv(out_dir / f'confmat_seed{run_seed}.csv', index=False)
    return test_acc, test_f1

# -------------------
# Main Function
# -------------------
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--n_runs', type=int, default=5)
    parser.add_argument('--n_samples', type=int, default=1000)
    parser.add_argument

### CODE REVIEW

**Comment by NaderMalki** [comment]
import torch
import torch.nn as nn
import torch.nn.functional as F

# ----------------------------
# تنظیمات و قراردادها (شکلِ داده)
# ----------------------------
# فرض: ورودی‌های زمان-سری با شکل (batch, seq_len, input_size)
# یا برای سیگنال‌های چندکاناله (batch, seq_len, n_channels).
# برای متن/کاتالوگ: می‌توان ابتدا embedding انجام داد تا به input_size برسد.
#
# پارامترهای پیشنهادی پیش‌فرض:
# input_size=128, seq_len=50, hidden_size=64, num_layers=2, n_classes=2
#
# تابع کمکی برای projection (در صورت نیاز به تغییر بعد ورودی):
class InputProjector(nn.Module):
    """
    پروجکت کردن ورودی به یک embedding ثابت (input_size_out).
    مفید برای ترکیب ویژگی‌های خام یا کاهش/افزایش بُعد.
    """
    def __init__(self, input_dim, output_dim, use_layernorm=False):
        super().__init__()
        self.fc = nn.Linear(input_dim, output_dim)
        self.use_ln = use_layernorm
        if use_layernorm:
            self.ln = nn.LayerNorm(output_dim)

**Comment by NaderMalki** [comment]
import torch
import torch.nn as nn
import torch.nn.functional as F

# ----------------------------
# تنظیمات و قراردادها (شکلِ داده)
# ----------------------------
# فرض: ورودی‌های زمان-سری با شکل (batch, seq_len, input_size)
# یا برای سیگنال‌های چندکاناله (batch, seq_len, n_channels).
# برای متن/کاتالوگ: می‌توان ابتدا embedding انجام داد تا به input_size برسد.
#
# پارامترهای پیشنهادی پیش‌فرض:
# input_size=128, seq_len=50, hidden_size=64, num_layers=2, n_classes=2
#
# تابع کمکی برای projection (در صورت نیاز به تغییر بعد ورودی):
class InputProjector(nn.Module):
    """
    پروجکت کردن ورودی به یک embedding ثابت (input_size_out).
    مفید برای ترکیب ویژگی‌های خام یا کاهش/افزایش بُعد.
    """
    def __init__(self, input_dim, output_dim, use_layernorm=False):
        super().__init__()
        self.fc = nn.Linear(input_dim, output_dim)
        self.use_ln = use_layernorm
        if use_layernorm:
            self.ln = nn.LayerNorm(output_dim)

    def forward(self, x):
        # x: (B, T, D_in)
        out = self.fc(x)
        if self.use_ln:
            out = self.ln(out)
        return F.relu(out)
